url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/layout/JGraphWordBrowser.java

```diff
@@ -0,0 +1,480 @@
+				newCells.add(sourceVertex);
+			}
+
+			// Creates the neighbour
+			Object targetVertex = mapping.get(target);
+			if (targetVertex == null) {
+				targetVertex = createCell(target, vertexStylename, false);
+				newCells.add(targetVertex);
+			}
+
+			// Creates the edge to the neighbour
+			Object edge = (data != null) ? mapping.get(data) : null;
+			ConnectionSet cs = null;
+			if (edge == null) {
+				edge = createCell(data, edgeStylename, true);
+				newCells.add(edge);
+				if (getChildCount(sourceVertex) > 0
+						&& getChildCount(targetVertex) > 0)
+					cs = new ConnectionSet(edge, getDefaultPort(sourceVertex),
+							getDefaultPort(targetVertex));
+			}
+
+			// Inserts the new cells into the graph model
+			if (!newCells.isEmpty()) {
+				insert(newCells.toArray(), null, cs, null, null);
+			}
+		}
+		return data;
+	}
+
+	/**
+	 * Uses the factory to create a vertex or edge for a specified stylename and
+	 * registers the mapping.
+	 */
+	protected Object createCell(Object data, String stylename, boolean isEdge) {
+		Object cell = (isEdge) ? createEdge(data) : createVertex(data);
+		if (data != null) {
+			mapping.put(data, cell);
+		}
+
+		// Stores the stylename in the attribute map
+		if (stylename != null) {
+			//Map attributes = getAttributes(cell);
+			// JSPGraphConstants.setStyle(attributes, stylename);
+		}
+
+		return cell;
+	}
+
+	/**
+	 * Hook for subclassers to create the vertex.
+	 */
+	public Object createVertex(Object data) {
+		DefaultGraphCell vertex = new DefaultGraphCell(data);
+		GraphConstants.setAutoSize(vertex.getAttributes(), true);
+		GraphConstants.setInset(vertex.getAttributes(), 4);
+		vertex.addPort();
+		return vertex;
+	}
+
+	/**
+	 * Hook for subclassers to create the vertex.
+	 */
+	public Object createEdge(Object data) {
+		DefaultEdge edge = new DefaultEdge(data);
+		int len = String.valueOf(data).length();
+		int size = (len > 20) ? ((len > 40) ? 8 : 10) : 12;
+		GraphConstants
+				.setFont(edge.getAttributes(), new Font("Serif", 0, size));
+		GraphConstants.setLabelAlongEdge(edge.getAttributes(), true);
+		GraphConstants.setLineColor(edge.getAttributes(), Color.GRAY);
+		// GraphConstants.setRouting(edge.getAttributes(),
+		// JSPGraphParallelEdgeRouter.sharedInstance);
+		return edge;
+	}
+
+	/**
+	 * Removes a cell from the model.
+	 */
+	public Object remove(Object data) {
+		Object vertex = mapping.remove(data);
+		if (vertex != null) {
+
+			// Includes all connections to/from the cell
+			Set cells = DefaultGraphModel.getEdges(this,
+					new Object[] { vertex });
+			cells.add(vertex);
+
+			// Removes the cells from the model
+			remove(cells.toArray());
+		}
+		return vertex;
+	}
+
+	/**
+	 * Returns the default port for the specified vertex.
+	 */
+	public Object getDefaultPort(Object vertex) {
+		return getChild(vertex, 0);
+	}
+
+	/**
+	 *
+	 * 
+	 */
+	public void add(String word, String wordType) {
+		if (word.endsWith(")")) {
+			int index = word.indexOf(" (");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 235 does not find the substring and returns -1, the call to substring() on line 237 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string word.